odoo/models.py

```diff
@@ -3427,7 +3427,17 @@ def unlink(self):
         self._check_concurrency()
 
         from odoo.addons.base.models.ir_model import MODULE_UNINSTALL_FLAG
-        for func in self._ondelete_methods:
+        # go through _ondelete_methods in reverse order:
+        # let module A be a module that implements an _ondelete_method X for model M
+        # let module B be a module that extends A's model M and introduces _ondelete_method Y
+        # let both conditions for X and Y be True and trigger the respective exceptions
+        # if module B introduces tests that check for Y's exception to be raised, and the
+        # record on which the test is being performed happens to also trigger X's exception (or
+        # any other number of exceptions implemented elsewhere that the developer of B is not aware
+        # of), then the test will fail because X's exception will trigger first and the test
+        # will never reach Y's exception. By reversing the order we guarantee that the exception
+        # raised (if multiple) is the one implemented by the last extending module.
+        for func in self._ondelete_methods[::-1]:
```

Comment:
`reversed` might be clearer there?

Reply:
ok but for the record, `[::-1]` is and will always be cooler

Reply:
alternatively a stack could be used for `_ondelete_methods` but I think it's overkill

Reply:
Yeah, the slicing trick is unreadable.

Reply:
> ok but for the record, `[::-1]` is and will always be cooler

![truth](https://user-images.githubusercontent.com/7571158/101030664-ac1f9d00-3579-11eb-9cc7-f1eba48d4ee0.png)